### PR TITLE
Work for release 0.21 review

### PR DESCRIPTION
**Pull request by sbrossie**

### CODE REVIEW

**Comment by sbrossie** [comment]
@pierre for review

File: invoice/src/main/java/org/killbill/billing/invoice/dao/InvoiceDaoHelper.java [hunk 1]
```diff
@@ -331,6 +334,32 @@ public boolean apply(final InvoiceModelDao input) {
         }
     }
 
+    private void setTrackingIdsFromTransaction(final Iterable<InvoiceModelDao> invoices, final EntitySqlDaoWrapperFactory entitySqlDaoWrapperFactory, final InternalTenantContext context) {
+        final InvoiceTrackingSqlDao invoiceTrackingidSqlDao = entitySqlDaoWrapperFactory.become(InvoiceTrackingSqlDao.class);
+        final List<InvoiceTrackingModelDao> trackingIds = invoiceTrackingidSqlDao.getByAccountRecordId(context);
+
+        final Map<UUID, List<InvoiceTrackingModelDao>> invoiceTrackingIdsPerInvoiceId = new HashMap<UUID, List<InvoiceTrackingModelDao>>();
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/dao/InvoiceDaoHelper.java]: A `MultiMap` might have helped here.

File: invoice/src/main/java/org/killbill/billing/invoice/dao/InvoiceDaoHelper.java [hunk 2]
```diff
@@ -331,6 +334,32 @@ public boolean apply(final InvoiceModelDao input) {
         }
     }
 
+    private void setTrackingIdsFromTransaction(final Iterable<InvoiceModelDao> invoices, final EntitySqlDaoWrapperFactory entitySqlDaoWrapperFactory, final InternalTenantContext context) {
+        final InvoiceTrackingSqlDao invoiceTrackingidSqlDao = entitySqlDaoWrapperFactory.become(InvoiceTrackingSqlDao.class);
+        final List<InvoiceTrackingModelDao> trackingIds = invoiceTrackingidSqlDao.getByAccountRecordId(context);
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/dao/InvoiceDaoHelper.java]: Not suggesting to change this for now, but a bit worried about memory usage...

File: invoice/src/main/java/org/killbill/billing/invoice/model/DefaultInvoice.java [hunk 3]
```diff
@@ -158,6 +161,17 @@ public boolean addInvoiceItems(final Collection<InvoiceItem> items) {
...
     }
 
+    @Override
+    public List<String> getTrackingIds() {
+        return trackingIds;
+    }
+
+    @Override
+    public boolean addTrackingIds(final Collection<String> trackingIds) {
+        this.trackingIds.addAll(trackingIds);
```
Comment (pierre) [inline invoice/src/main/java/org/killbill/billing/invoice/model/DefaultInvoice.java]: You could have returned the result of `addAll` to make the returned value a bit more meaningful. This (historic) boolean API is a bit weird though.

File: usage/src/main/java/org/killbill/billing/usage/api/user/DefaultUsageUserApi.java [hunk 4]
```diff
@@ -67,15 +68,23 @@ public DefaultUsageUserApi(final RolledUpUsageDao rolledUpUsageDao,
...
+            trackingIds = UUIDs.randomUUID().toString();
         // check if we have (at least) one row with the supplied tracking id
-        if (!Strings.isNullOrEmpty(record.getTrackingId()) && recordsWithTrackingIdExist(record, internalCallContext)) {
+        } else if (recordsWithTrackingIdExist(record, internalCallContext)) {
             throw new UsageApiException(ErrorCode.USAGE_RECORD_TRACKING_ID_ALREADY_EXISTS, record.getTrackingId());
+        } else {
+            trackingIds = record.getTrackingId();
+        }
+
+        if (!Strings.isNullOrEmpty(record.getTrackingId()) && recordsWithTrackingIdExist(record, internalCallContext)) {
```
Comment (pierre) [inline usage/src/main/java/org/killbill/billing/usage/api/user/DefaultUsageUserApi.java]: Empty `if` block?

File: invoice/src/main/java/org/killbill/billing/invoice/dao/InvoiceDaoHelper.java (hunk 2 above)
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/dao/InvoiceDaoHelper.java]: Ack - will pull info for the matching invoices.

File: invoice/src/main/java/org/killbill/billing/invoice/model/DefaultInvoice.java (hunk 3 above)
Comment (sbrossie) [reply invoice/src/main/java/org/killbill/billing/invoice/model/DefaultInvoice.java]: Ack - will fix

File: usage/src/main/java/org/killbill/billing/usage/api/user/DefaultUsageUserApi.java (hunk 4 above)
Comment (sbrossie) [reply usage/src/main/java/org/killbill/billing/usage/api/user/DefaultUsageUserApi.java]: Yep, mistake will mean

**Pull request merged**